metapool/metapool.py

```diff
@@ -480,6 +480,147 @@ def read_pico_csv(f, sep='\t', plate_reader='SpectraMax_i3x',
+    conc_dict : dict
+        Dictionary containing suffixes and filepaths for concentration file(s)
+    plate_reader : str
+        Plate reader used to generate quant file(s); see read_pico_csv for
+        supported values.  Must be the same for all files.
+
+    Returns:
+    a_df : pandas DataFrame
+        An expansion of the input plate_df, with one concentration column added
+        for each entry in conc_dict.
+    """
+
+    a_df = plate_df.copy()
+    for curr_dilution_suffix, curr_conc_fp in conc_dict.items():
+        curr_conc_df = _read_and_label_pico_csv(
+            curr_conc_fp, curr_dilution_suffix, plate_reader=plate_reader)
+        a_df = pd.merge(a_df, curr_conc_df, on=PM_WELL_KEY)
+    # next dilution suffix
+    return a_df
+
+
+def select_sample_dilutions(
+        a_plate_df, dilution_suffixes_in_order, a_mask_func,
+        a_col_name_func=None):
+    """
+    Selects the correct source of each sample based on input dilutions
+
+    Parameters:
+    a_plate_df : pandas DataFrame
+        A DataFrame of plate info, containing at least PM_WELL_KEY,
+        PM_PROJECT_PLATE_KEY, and PM_COMPRESSED_PLATE_NAME_KEY columns
+    dilution_suffixes_in_order : list of str
+        List of dilution suffix(es) in order from most to least preferred for
+        use. NOTE THAT the suffix for the base dilution plate must be the
+        LAST ONE.  The base dilution plate doesn't have to be *undiluted*,
+        but it must be the one that will be used when no other dilution meets
+        the user's criteria.
+    a_mask_func : function
+        A function to generate a mask that is true for samples that meet the
+        user's (arbitrary) criteria, taking as parameters a DataFrame and a
+        particular dilution's (full) concentration column name.
+    a_col_name_func : function
+        A function to generate the (existing) column name of a concentration
+        column in the plate dataframe based on its dilution suffix. If None,
+        defaults to SAMPLE_DNA_CONC_KEY_<suffix>, which matches what is
+        produced by _read_and_label_pico_csv.
+
+    Returns:
+    a_plate_df : pandas DataFrame
+        An expansion of the input plate_df; for each sample, the concentration
+        column is set to the sample's concentration in the most preferred
+        concentration that meets the user's criteria. Likewise, the plate name
+        and project plate columns are set to those for the selected dilution,
+        and the 'Diluted' column is set to true if the selected dilution is
+        NOT the base dilution.
+    """
+
+    if PM_DILUTED_KEY in a_plate_df.columns:
+        # function has already been applied to data
+        warnings.warn('Dilution operation was already performed')
+        return a_plate_df
+
+    def _default_col_name_func(x):
+        return f"{SAMPLE_DNA_CONC_KEY}_{x}"
+
+    a_col_name_func = _default_col_name_func if a_col_name_func is None \
+        else a_col_name_func
```

Comment:
Out of curiosity, how ofter we would expect the calls to `select_sample_dilutions` have a_col_name_func as None? It took me a second to understand what's going on here and I'm not sure this is necessary.

Reply:
Ok, I'm willing to take it out :)